modules/extension/wmts/src/main/java/org/geotools/data/wmts/request/AbstractGetTileRequest.java

```diff
@@ -0,0 +1,401 @@
+//        }
+
+        LOGGER.warning("===== getTiles    layer:" + layer + " w:" + requestedWidth + " x h:"+requestedHeight);
+
+//        int w = Integer.parseInt(width);
+//        int h = Integer.parseInt(height);
+        TileMatrixSet matrixSet = selectMatrixSet();
+        
+        String requestUrl = onlineResource.toString();
+        if (WMTSServiceType.REST.equals(type)) {
+            String format = (String) getProperties().get("Format");
+            if (format == null || format.isEmpty()) {
+                format = "image/png";
+                LOGGER.fine("Format not set, trying with " + format);
+            }
+            requestUrl = layer.getTemplate(format);
+            if(requestUrl == null) {
+                LOGGER.info("Template URL not available for format  " + format);
+                format = layer.getFormats().get(0);
+                if(LOGGER.isLoggable(Level.FINE)) {
+                    LOGGER.fine("Available formats: " + layer.getFormats() + " -- Selecting " + format);
+                }
+                requestUrl = layer.getTemplate(format);
+            }
+        }
+
+        WMTSTileService wmtsService = new WMTSTileService(requestUrl, type, layer, styleString, matrixSet);
+
+        wmtsService.getDimensions().put(WMTSTileService.DIMENSION_TIME, requestedTime);
+
+        ((Map)(wmtsService.getExtrainfo()
+                .computeIfAbsent(WMTSTileService.EXTRA_HEADERS, extra -> new HashMap<>())
+                )).putAll(this.headers);
+
+        // zoomLevel = factory.getZoomLevel(zoom, wmtsService);
+        int scale = 0;
+
+        try {
+            scale = (int) Math.round(RendererUtilities.calculateScale(requestedBBox, requestedWidth, requestedHeight, DPI));
+        } catch (FactoryException | TransformException ex) {
+            throw new RuntimeException("Failed to calculate scale", ex);
+        }
+        tiles = wmtsService.findTilesInExtent(requestedBBox, scale, false, MAXTILES);
+        LOGGER.fine("found " + tiles.size() + " tiles in " + requestedBBox);
+        if (tiles.isEmpty()) {
+            return tiles;
+        }
+        Tile first = tiles.iterator().next();
+        int z = first.getTileIdentifier().getZ();
+        List<TileMatrixLimits> limits = layer.getTileMatrixLinks().get(matrixSet.getIdentifier())
+                .getLimits();
+        TileMatrixLimits limit;
+        if (!limits.isEmpty()) {
+
+            limit = limits.get(z);
+        } else {
+            // seems that MapProxy (and all REST APIs?) don't create limits
+            limit = new TileMatrixLimits();
+            limit.setMaxCol(matrixSet.getMatrices().get(z).getMatrixWidth() - 1);
+            limit.setMaxRow(matrixSet.getMatrices().get(z).getMatrixHeight() - 1);
+            limit.setMinCol(0);
+            limit.setMinRow(0);
+            limit.setTileMatix(matrixSet.getIdentifier());
+        }
+        ArrayList<Tile> remove = new ArrayList<>();
+        for (Tile tile : tiles) {
```

Comment:
The initial set of tiles is only filtered according to the tilematrix and the requested bbox.
The layer may have further limits, so we are removing tiles which do not contain layer data.
No data is fetched here: we are providing the set of tiles that shall be retrieved in a next step.